quinn-h3/src/proto/headers.rs

```diff
@@ -0,0 +1,296 @@
+                Field::Authority(a) => pseudo.authority = Some(a),
+                Field::Path(p) => pseudo.path = Some(p),
+                Field::Status(s) => pseudo.status = Some(s),
+                Field::Header((n, v)) => {
+                    fields.append(n, v);
+                }
+            }
+        }
+
+        Ok(Header { pseudo, fields })
+    }
+}
+
+enum Field {
+    Method(Method),
+    Scheme(String<Bytes>),
+    Authority(String<Bytes>),
+    Path(String<Bytes>),
+    Status(StatusCode),
+    Header((HeaderName, HeaderValue)),
+}
+
+impl Field {
+    pub fn parse<N, V>(name: N, value: V) -> Result<Self, Error>
+    where
+        N: AsRef<[u8]>,
+        V: AsRef<[u8]>,
+    {
+        let name = name.as_ref();
+        if name.len() < 1 {
+            return Err(Error::InvalidHeaderName(format!("{:?}", name)));
+        }
+
+        if name[0] == b':' {
+            let pseudo = match PSEUDO_MAP.get(name) {
+                Some(pseudo) => pseudo,
+                None => return Err(Error::invalid_name(name)),
+            };
+            return Ok(match pseudo {
+                PseudoType::SCHEME => Field::Scheme(try_value(name, value)?),
+                PseudoType::AUTHORITY => Field::Authority(try_value(name, value)?),
+                PseudoType::PATH => Field::Path(try_value(name, value)?),
+                PseudoType::METHOD => Field::Method(
+                    Method::from_bytes(value.as_ref())
+                        .or(Err(Error::invalid_value(name, value)))?,
+                ),
+                PseudoType::STATUS => Field::Status(
+                    StatusCode::from_bytes(value.as_ref().into())
+                        .or(Err(Error::invalid_value(name, value)))?,
+                ),
+            });
+        }
+
+        Ok(Field::Header((
+            HeaderName::from_bytes(name.as_ref().into()).or(Err(Error::invalid_name(name)))?,
+            HeaderValue::from_bytes(value.as_ref().into())
+                .or(Err(Error::invalid_value(name, value)))?,
+        )))
+    }
+}
+
+fn try_value<N, V>(name: N, value: V) -> Result<String<Bytes>, Error>
+where
+    N: AsRef<[u8]>,
+    V: AsRef<[u8]>,
+{
+    string::TryFrom::<Bytes>::try_from(Bytes::from(value.as_ref()))
+        .or(Err(Error::invalid_value(name, value)))
+}
+
+macro_rules! pseudo_type {
+    (
+        $(
+            ($name:ident, $val:expr),
+        )+
+    ) => {
+        #[derive(Clone)]
+        enum PseudoType { $($name,)* }
+
+        lazy_static! {
+            static ref PSEUDO_MAP: HashMap<Cow<'static, [u8]>, PseudoType> = [
+                $((Cow::Borrowed(&$val[..]), PseudoType::$name),)+
+            ].into_iter().map(|(n, v)| (n.clone(), v.clone())).collect();
+        }
+    }
+}
+
+pseudo_type![
+    (METHOD, b":method"),
+    (SCHEME, b":scheme"),
+    (AUTHORITY, b":authority"),
+    (PATH, b":path"),
+    (STATUS, b":status"),
+];
+
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct Pseudo {
```

Comment:
This whole term of the "pseudo-header" isn't something I'm familiar with. Maybe add a few lines of documentation to clarify what/why it's here?

Reply:
done in f869a8e